src/tools/perf/lib/libperf.c

```diff
@@ -1279,25 +1377,33 @@ static ucs_status_t ucp_perf_test_setup_endpoints(ucx_perf_context_t *perf,
         return UCS_ERR_UNSUPPORTED;
     }
 
-    /* pack the local endpoints data and send to the remote peer */
+    /* Pack the local endpoints data and send to the remote peer */
     status = ucp_perf_test_send_local_data(perf, features);
     if (status != UCS_OK) {
         goto err;
     }
 
-    /* receive remote peer's endpoints' data and connect to them */
+    /* Receive remote peer's endpoints' data and connect to them */
     status = ucp_perf_test_receive_remote_data(perf, peer_index);
     if (status != UCS_OK) {
         goto err;
     }
 
-    /* sync status across all processes */
+    /* Sync status across all processes */
     status = ucx_perf_test_exchange_status(perf, UCS_OK);
     if (status != UCS_OK) {
         goto err_destroy_eps;
     }
 
-    /* force wireup completion */
+    /* Create self_ep used for copying into non-host memory types*/
+    if (!ucp_perf_test_is_host_mem_only(&perf->params)) {
+        status = ucp_perf_test_setup_self_endpoints(perf);
+        if (status != UCS_OK) {
+            goto err_destroy_eps;
```

Comment:
`ucp_perf_test_destroy_eps` will be called twice in case of error inside `ucp_perf_test_setup_self_endpoints`.
Please split clean-up of ep, rkey, and self_ep, self_send_rkey, self_recv_rkey into two methods.